eevee/metrics/entity.py

```diff
@@ -19,21 +44,80 @@ def entity_report(true_labels: pd.DataFrame, pred_labels: pd.DataFrame) -> pd.Da
+    df["true_ent_type"] = df["true"].apply(lambda it: it[0].get("type"))
+    df["pred_ent_type"] = df["pred"].apply(lambda it: it[0].get("type"))
+
+    df["true_ent_value"] = df["true"].apply(lambda it: it[0]["values"][0].get("value"))
+    df["pred_ent_value"] = df["pred"].apply(lambda it: it[0]["values"][0].get("value"))
+
     # All the unique entity types in the dataset
     entity_types = sorted(set([ent["type"] for ent in py_.flatten(df["true"].tolist() + df["pred"].tolist())]))
 
     # TODO: Handle compositional entities like datetime
     report = []
 
     for entity_type in entity_types:
-        report.append({
-            "Entity": entity_type,
-            "FPR": "NA",
-            "FNR": "NA",
-            "Mismatch Rate": "NA",
-            "Support": "NA",
-            "Positives": "NA",
-            "Negatives": "NA"
-        })
+
+        if entity_type in ENTITY_EQ_FNS:
+
+            entity_type_df = df[(df["true_ent_type"] == entity_type) | (df["pred_ent_type"] == entity_type)]
+
+            y_true = []
+            y_pred = []
+
+            y_true_mmr = []
+            y_pred_mmr = []
+
+            eq_fn_for_this_entity = ENTITY_EQ_FNS[entity_type]
+
+            for _, row in entity_type_df.iterrows():
+
+                true_ent = row["true"][0]
+                pred_ent = row["pred"][0]
+
+                if row["true_ent_type"] == entity_type and row["pred_ent_type"] == entity_type:
+
+                    is_this_entity_type_value_equal = eq_fn_for_this_entity(true_ent, pred_ent)
+                    if is_this_entity_type_value_equal:
+                        y_true.append(True)
+                        y_pred.append(True)
+
+                # false negative, we expected a prediction but didn't happen.
+                elif row["true_ent_type"] == entity_type and row["pred_ent_type"] != entity_type:
+                    y_true.append(True)
+                    y_pred.append(None)
+
+                # false positive, no prediction should have happened
+                elif row["true_ent_type"] != entity_type and row["pred_ent_type"] == entity_type:
+                    y_true.append(None)
+                    y_pred.append(True)
+
+                # else:
+                #     # TODO: don't know how to handle. types matching, but values not matching.
+                #     # that is mismatch rate?
+                #     pass
+                #     # y_true.append(None)
+                #     # y_pred.append(None)
+
+                y_true_mmr.append(row["true"][0])
+                y_pred_mmr.append(row["pred"][0])
+
+            ent_fpr = slot_fpr(y_true, y_pred)
+            ent_fnr = slot_fnr(y_true, y_pred)
+            ent_mmr = mismatch_rate(y_true_mmr, y_pred_mmr)
+
+            ent_pos = slot_positives(y_true, y_pred)
+            ent_neg = slot_negatives(y_true, y_pred)
+
+            report.append({
+                "Entity": entity_type,
+                "FPR": f"{ent_fpr:.2f}",
```

Comment:
don't put formatted strings, this is supposed to be true value

Reply:
oh okay, i'll leave them as floats then.